server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -312,6 +312,10 @@ void maybePutIndexUnderLock(BytesRef uid, IndexVersionValue version) {
         if (maps.isSafeAccessMode()) {
             putIndexUnderLock(uid, version);
         } else {
+            // The existing delete tombstone is no longer the latest version (.i.e stale)
+            // when a newer version is processed regardless of the safe mode of the version map.
+            // Therefore, it's better to always prune that tombstone to avoid accidental accesses.
```

Comment:
How about:
// Even though we don't store a record of the indexing operation (and mark as unsafe) we should still remove any previous delete for this uuid. Note that this is expected to not hurt performance because the tombstone is small (if not empty) when usafe is relevant.

Reply:
I've updated the comment.